src/annotator/anchoring/types.js

```diff
@@ -49,35 +42,55 @@ export class RangeAnchor {
   /**
    * Create an anchor from a serialized `RangeSelector` selector.
    *
-   * @param {Node} root -  A root element from which to anchor.
+   * @param {Element} root -  A root element from which to anchor.
    * @param {RangeSelector} selector
    */
   static fromSelector(root, selector) {
-    const data = {
-      start: selector.startContainer,
-      startOffset: selector.startOffset,
-      end: selector.endContainer,
-      endOffset: selector.endOffset,
-    };
-    const range = new SerializedRange(data);
+    const startContainer = nodeFromXPath(selector.startContainer, root);
+    if (!startContainer) {
+      throw new Error('Failed to resolve startContainer XPath');
+    }
+
+    const endContainer = nodeFromXPath(selector.endContainer, root);
+    if (!endContainer) {
+      throw new Error('Failed to resolve endContainer XPath');
+    }
+
+    const startPos = TextPosition.fromCharOffset(
+      startContainer,
+      selector.startOffset
+    );
+    const endPos = TextPosition.fromCharOffset(
+      endContainer,
+      selector.endOffset
+    );
+
+    const range = new TextRange(startPos, endPos).toRange();
     return new RangeAnchor(root, range);
   }
 
   toRange() {
-    return this.range.toRange();
+    return this.range;
   }
 
   /**
    * @return {RangeSelector}
    */
   toSelector() {
-    const range = this.range.serialize(this.root);
+    // "Shrink" the range so that it tightly wraps its text. This ensures more
+    // predictable output for a given text selection.
+    const normalizedRange = TextRange.fromRange(this.range).toRange();
```

Comment:
Given an HTML structure like:

```html
<div>Foo bar<b></b></div>
```

If a user selects the text "Foo bar" then the browser's selection (`window.getSelection().getRangeAt(0)`) could represent that in several different ways (eg. it could select the whole `<div>`, just the text "Foo bar", or the text "Foo bar" plus the empty `<b>` tag). This normalization results in a range that tightly wraps the text.